### PR TITLE
Add Spines class as a container for all Axes spines

### PR DESCRIPTION
**Pull request by timhoffm**
## PR Summary

Closes #17099. Alternative approach to #17100, as proposed in https://github.com/matplotlib/matplotlib/pull/17099#issuecomment-612603149.

With this, you can do (updated based on discussion below):

~~~
spines['top'].set_visible(False)              # as before
spines.top.set_visible(False)                 # NEW: access single spines by attibute
spines[:].set_visible(False)                 # NEW: batch-set properties for all spines
spines[['top', 'right']].set_visible(False)  # NEW: batch-set properties for some spines
~~~

and it should be mostly backward-compatible (with the exception that I do not support OrderedDict-specific methods) for `spines`.

*Note*: This PR consists of two patches for easier review: The first patch introduces the new API, the second patch applies it throughout the library (i.e. you can see the API in action here).

Open issues:

- [ ] needs discussion
  - [ ] is it generally a good idea to have a collection object instead of `OrderedDict`?
  - [x] do we want access by attribute (`spines.top.set_visible`)? - Access by item `spines['top']` will stay anyway for backward-compatibility
  - [x] do we want a proxy for changing all `spines.all.set_visible(False)`?
    This is now implemented as `spines[:].set_visible(False)`.
  - [ ] is it ok to be not fully backward compatible by just deriving from `MutableMapping`? - I.e. the `OrderedDict` methods `spines.popitem()` and `spines.move_to_end()` are dropped without deprecation. - Could add deprecation dance for these as well if really needed.
  - [x] *Currently not implemented*: Do we want to support subgroups `spines['left', 'right']` or (personally preferred) `spines[['left', 'right']]`?
    Yes. Implemented.
  - [ ] *Currently not implemented*: Do we want to deprecate modifying the available spines (i.e. deprecate towards using a `Mapping` instead of a `MutableMapping`.
- [x] compare with AxisDict approach (see https://github.com/matplotlib/matplotlib/pull/17107#issuecomment-612666399)
- [x] pickle support
- [x] use `spines.all` throught library and examples where suitable
- [x] what's new entry

### CODE REVIEW

**Comment by anntzer** [comment]
Haven't reviewed carefully, but you may want to compare with the AxisDict approach in mpl_toolkits, which similarly overrides getitem so that one can write `ax.axis["left", "top"].toggle(...)` or `ax.axis[:].toggle(...)` (`.axis` is "similar" to the spines dict).

**Comment by timhoffm** [comment]
### Comparison with `AxisDict` approach

- `AxesDict` only supports item access (single item `axis['top']` and all items via slice `axis[:]`).
  It's an intended feature that `Spines` makes the Elements available via attribute access (`spines.top`). It also has item access, but rather for backward compatibility. It currently does not support accessing all elments via a full slice, because I rather see item access as backward compatibility and we currently don't have hat. But it would be easy to add if desired.
- The multi-element proxy object returned by `AxisDict` is `SimpleChainedObjects`, which has a simpler implementation than the `SpinesProxy`. However, that means it misses some features of SpinesProxy:
  - `SpinesProxy` limits the available methods to `set_*` methods. Other methods are most likely not reasonable to be batch-called. In particular all methods mainly returning a value are not supported (`SimpleChainedObjects` still allows to call these methods but does not return anything).
    With the limitation I'd rather err on the side of being too restrictive than to allow calling nonsensical methods. It's always possible to whitelist additional methods later if needed.
  - `SpinesProxy` has a somewhat reasonable docstring. `SimpleChainedObjects` does not and it's more hard to describe, because it's recursive: In `axis[:].set_visible(False)`, `axis[:]` is a `SimpleChainedObjects` containing spines, and `axis[:].set_visible` is a `SimpleChainedObject` containing bound `Spine.set_visible` methods.
    In contrast `spines.all.set_visible` is a dynamically created function with the correct docstring.
- `AxisDict` is a `dict` subclass, which may expose undesired `dict` functionality: Should the user be able do modify an `AxisDict` by adding elements, removing elements or replacing elements? I say rather not. While `Spines` currently allows that for backward compatibility, I'm considering deprecating these and finally making `Spines` read-only by deriving from `Mapping` instead of `MutableMapping`.


**Summary:** `Spines` + `SpineProxy` intentionally chooses a slightly different design compared to `AxesDict` + `SimpleChainedObjects` so that it provides a more well tailored interface due to:
- proper docstrings
- support for tab-completion
- limiting to broadcasting setter methods

**Comment by anntzer** [comment]
We don't necessarily have to copy the AxesDict approach here.  However, I think we're also trying to bring axes_grid towards inclusion in the main lib (https://github.com/matplotlib/matplotlib/issues/17095), so I guess the converse question could be, would it make sense to change AxesDict to use this machinery (with the proper deprecation period, yada yada)?  (You don't have to write the adaptor API for this PR, but it would be better to at least check something similar can be used for axes_grid.)

**Comment by timhoffm** [comment]
Without having written the prove of concept: From what I've seen in the comparison to `AxisDict` (https://github.com/matplotlib/matplotlib/pull/17107#issuecomment-612666399), I assume that it's quite easy to replace `AxisDict` with this approach.

**Review by QuLogic — DISMISSED** [review]
I don't have any changes to request, but I expect this would have to be checked against Cartopy, WCSAxes, and maybe yt, so blocking just in case.

**Comment by timhoffm** [comment]
**Cartopy:**
- uses `spines['geo']` (read/write) and `spines.values()`
- overrides `_gen_axes_spines()`
all of these usages are supported by this PR. - I've run the Cartopy tests: the 4 same tests failed with and without this PR.

**WCSA Axes:**
I didn't find any `spines` usage. All WCSA tests pass with this PR.

**yt:**
I didn't find any `spines` usage. (I did not bother to install a dev version to run the tests.)

**Comment by jklymak** [comment]
This seems pretty nice to me.  I assume users can add more spines as before.  How do we get the list of spines?  `ax.spines.keys()`?  I don't find the `abc` documentation to be super clear.

**Comment by timhoffm** [comment]
The list of spines is still `ax.spines.values()`. `axes.spines.keys()` would give you "top, "bottom", ...

By subclassing `MutableMapping` and implementing its abstract methods, we get a class with essentially the same API as a dict. Therefore `Spines` is almost a fully compatible replacement for the original `OrderedDict` (the slight incompatibility coming from that I did not bother to implement the special `OrderedDict` methods `popitem()` and `move_to_end()`).

**Comment by timhoffm** [comment]
Ping @QuLogic https://github.com/matplotlib/matplotlib/pull/17107#issuecomment-614915828 addresses your change request.

File: lib/matplotlib/spines.py [hunk 1]
```diff
@@ -534,3 +537,97 @@ def set_color(self, c):
...
+    the subset of spines that support it.
+    """
+    def __init__(self, spine_dict):
+        self._spine_dict = spine_dict
+
+    def __getattr__(self, name):
+        broadcast_targets = [spine for spine in self._spine_dict.values()
+                             if hasattr(spine, name)]
+        if not name.startswith('set_') or not broadcast_targets:
+            raise ValueError(f"'SpinesProxy' object has no attribute '{name}'")
```
Comment (QuLogic) [inline lib/matplotlib/spines.py:557]: Wouldn't that be `AttributeError`, then?

File: lib/matplotlib/spines.py [hunk 2]
```diff
@@ -534,3 +537,98 @@ def set_color(self, c):
...
+
+    def __setstate__(self, state):
+        self.__init__(**state)
+
+    def __getattr__(self, name):
+        try:
+            return self._dict[name]
+        except KeyError:
+            raise ValueError(f"'Spines' object does not contain a '{name}' "
+                             "spine")
```
Comment (QuLogic) [inline lib/matplotlib/spines.py:617]: Might as well join these strings.
```suggestion
            raise ValueError(
                f"'Spines' object does not contain a '{name}' spine")
```

**Review by QuLogic — APPROVED** [review]

**Comment by anntzer** [comment]
On thing I had missed is that in https://github.com/matplotlib/matplotlib/pull/17107#issuecomment-612666399 @timhoffm stated
> AxesDict only supports item access (single item axis['top'] and all items via slice axis[:]).

but that is actually missing an important point about AxesDict which is not provided by SpinesProxy: you can actually also select a subset of Axes using AxesDict, using e.g. `axis["top", "right"]` (which returns a proxy object on which you can call normal axes methods).

I'm not saying we absolutely need this feature, but would at least like to point out that this is something much easier to retrofit when using indexing notation rather than attribute access.

**Comment by timhoffm** [comment]
I agree that we should support subsets, but with `spines[['left', 'right']].

Thinking a bit more, I take the standpoint that we should support
~~~
# Single spine
spines['left'],   spines.left

# All spines
spines[:],        spines.all

# Subset of spines
spines[['left', 'right']]
~~~
This is an API similar to pandas and thus hopefully easy to remember. Differences are: 1) We have `.all` for convenience. 2) We don't have arbitrary slices because we don't have oder.

**Comment by anntzer** [comment]
Do we want `spines.all`, or is `spines[:]` good enough?
Also note that if you support `spines[["left", "right"]]`, then you'll likely end up supporting `spines["left", "right"]` as well unless you're being extra careful (the former calls `spines.__getitem__(["left", "right"])` and the latter `spines.__getitem__(("left", "rigtht"))`, the difference is just list/tuple).

**Comment by timhoffm** [comment]
One of the main reasons for this was to be able to access the spines via attributes (`spines.left` instead of `spines['left']`). In that way, it makes sense to have `spines.all`. The item access was originally intended only for backward-compatibility. However, when supporting subsets of spines, it's the only way to define them, which makes item-access first-class citizen again. In that sense, I agree, we don't need all. Btw. by removing `.all` this makes the API more pandas like, which IMHO is a good thing.

**Comment by QuLogic** [comment]
I'm a bit confused whether you do or do not want to keep `.all` now. At the very least, the doc note still mentions it and doesn't mention `[:]` that I can see.

**Comment by timhoffm** [comment]
Oh, will fix that. .all is gone.

File: lib/matplotlib/spines.py [hunk 3]
```diff
@@ -534,3 +537,114 @@ def set_color(self, c):
...
+        spines[['top', 'right']].set_visible(False)
+
+    Use an open slice to address all spines::
+
+        spines[:].set_visible(False)
+
+    """
+    def __init__(self, **kwargs):
+        self._dict = kwargs
+        self.all = SpinesProxy(self._dict)
```
Comment (QuLogic) [inline lib/matplotlib/spines.py:601]: Then this should be removed.

File: lib/matplotlib/spines.py [hunk 4]
```diff
@@ -534,3 +537,114 @@ def set_color(self, c):
...
+
+    The interface is dict-like mapping names (e.g. 'left') to `.Spine` objects.
+    Additionally it implements some pandas.Series-like features like accessing
+    elements by attribute::
+
+        spines['top'].set_visible(False)
+        spines.top.set_visible(False)
+
+    Multiple spines can be addressed simultaneously by passing a list. This
+    will return a `SpinesProxy` that broadcasts all ``set_*`` calls to it's
```
Comment (QuLogic) [inline lib/matplotlib/spines.py:589]:
```suggestion
    will return a `SpinesProxy` that broadcasts all ``set_*`` calls to its
```
but also, it should be made clear that _only_ `set_*` is broadcast.

**Review by efiring — APPROVED** [review]
This is very nice.
You could use the `ax.spines[['top', 'bottom']].set_visible(False)` type of construct many more places in the examples, but I certainly wouldn't insist on it.  It aids readability and reduces LOC whenever a setting is applied to more than one spine.

**Comment by timhoffm** [comment]
This has two positive reviews. IMHO it could be merged, or is there still anything to be discussed?

**Pull request merged**